### PR TITLE
HA mailbox, redolog

### PR DESCRIPTION
**Pull request by spoon16**
To enable a highly available (or horizontally scalable) mailbox we have eliminated as much local state management as possible. The mailbox redolog is an important piece of local state that needs to be maintained. It enabled account recovery and rollback, it is important independent of a highly available database.

Legacy mailbox workers wrote all redoable operations to a local redolog and all operations related to a specific mailbox would be captured on the same node. In a highly available mailbox worker pool redoable operations not all mailbox operations will be captured in the same redolog. Zimbra needs a distributed or shared redolog that facilitates capturing redoable operations across a mailbox worker pool.

Legacy 👴 mailbox workers wrote all redoable operations to a local redolog in an append only manner.
<p align="center">
  legacy mailbox, local redolog
</p>

Redoable operations are written during `com.zimbra.cs.mailbox.Mailbox.MailboxTransaction#close`. During the life cycle of a `MailboxTransaction` means that ordering will be implicitly handled by distributed locking (see #487). The main challenge after solving coordination is reliably being able to write data. We chose to use the mailbox DB as an intermediary, we've added a table to the DB and we write all redoable operations to the DB table (see Zimbra/zm-db-conf#2). Once the data is ordered properly in the database many different strategies can be used to extract that data and write to a redolog file on a networked file system. Some options include DB triggers or scheduled tasks (cron jobs). [Here is a Python script that demonstrates how to read the data from the DB and produce a valid redolog file.](https://gist.github.com/spoon16/abc5abed983470e51f96b9695005d6fa)

In a highly available 🔥 mailbox worker pool all redoable operations need to be captured and ordered across the mailbox worker pool (1 or many mailbox workers).
<p align="center">
  ha mailbox, distributed redolog
</p>

#### Related PRs
- Zimbra/zm-db-conf#2
- Zimbra/zm-build#76

### CODE REVIEW

File: store/src/java/com/zimbra/cs/redolog/RedoLogManager.java [hunk 1]
```diff
@@ -210,11 +218,20 @@ public synchronized void start() {
         }
 
         long fsyncInterval = RedoConfig.redoLogFsyncIntervalMS();
-        mLogWriter = createLogWriter(this, mLogFile, fsyncInterval);
+        if (useDbAsLogStorage){
+            mLogWriter = createLogWriter(this);
+        } else {
+            mLogWriter = mLogWriter = createLogWriter(this, mLogFile, fsyncInterval);
+        }
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/redolog/RedoLogManager.java]: Replace this with `mLogWriter = new DbLogWriter(this);`. Synacor can update to switch using dependency injection or configuration.

If dependency injection is eventually used the constructor for `DbLogWriter` and `FileLogWriter` should be updated to read whatever they need from the `RedoLogManager` passed to their constructor.

`FileLogWriter` constructor could be refactored to match `DbLogWriter` easily. We do not need to do this as part of this PR.

```java
    public FileLogWriter(RedoLogManager redoLogMgr) {
        mRedoLogMgr = redoLogMgr;

        mHeader = new FileHeader(sServerId);
        mFile = redoLogMgr.getLogFile();
        mFileSize = mFile.length();
        mLastLogTime = mFile.lastModified();

        mFsyncIntervalMS = RedoConfig.redoLogFsyncIntervalMS();
        mFsyncDisabled = DebugConfig.disableRedoLogFsync;

        mFsyncCount = mLogCount = 0;

        mCommitNotifyQueue = new CommitNotifyQueue(100);
    }
```

File: store/src/java/com/zimbra/cs/redolog/RedoLogManager.java [hunk 2]
```diff
@@ -354,7 +373,10 @@ public synchronized void stop() {
         }
 
         try {
-            forceRollover();
+            // rollover only is needed when use FileLogWriter mechanism
+            if (getLogWriter() instanceof FileLogWriter) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/redolog/RedoLogManager.java]: Change to `!(mLogWriter instanceof DbLogWriter)`. Use negation to ensure that we are changing the behavior for only the new `DbLogWriter`.

File: store/src/java/com/zimbra/cs/redolog/RedoLogManager.java [hunk 3]
```diff
@@ -380,7 +402,8 @@ public void log(RedoableOp op, boolean synchronous) {
 
         logOnly(op, synchronous);
 
-        if (isRolloverNeeded(false))
+        //rollover is needed only when use FileLogWriter mechanism
+        if (isRolloverNeeded(false) && (getLogWriter() instanceof FileLogWriter))
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/redolog/RedoLogManager.java]: Change to `!(mLogWriter instanceof DbLogWriter)`. Use negation to ensure that we are changing the behavior for only the new `DbLogWriter`.

File: store/src/java/com/zimbra/cs/redolog/RedoLogManager.java [hunk 4]
```diff
@@ -52,6 +53,9 @@
  */
 public class RedoLogManager {
 
+    // if useDbAsLogStorage eq true then will be used a DbLogWriter, FileLogWriter will be used otherwise
+    private boolean useDbAsLogStorage = true;
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/redolog/RedoLogManager.java]: Remove this boolean. Synacor will want to switch based on some configuration key and this complicates the future implementation of configuration based dependency injection.

Comment (jnambo) [reply store/src/java/com/zimbra/cs/redolog/RedoLogManager.java]: ok sure not problem

File: store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+    @Override
+    public synchronized long getSize() {
+        if (conn == null) {
+            throw new RuntimeException("Redolog connection closed");
+        }
+
+        try {
+            return DbDistibutedRedolog.getAllOpSize(conn);
+        } catch (ServiceException e) {
+            throw new RuntimeException(String.format(FAILED_METHOD_TEMPLATE, "getSize"), e);
```
[the remark this answers is not in the document]
Comment (jnambo) [reply store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java]: This Exception has to be thrown this way if we don't want to change the semantic of this method in the interface which we are implemented for.

File: store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+    public boolean delete() {
+        if (conn == null) {
+            throw new RuntimeException("Redolog connection closed");
+        }
+
+        try {
+            DbDistibutedRedolog.clearRedolog(conn);
+            return true;
+        } catch (ServiceException e) {
+            throw new RuntimeException(String.format(FAILED_METHOD_TEMPLATE, "delete"), e);
```
[the remark this answers is not in the document]
Comment (jnambo) [reply store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java]: This Exception has to be thrown this way if we don't want to change the semantic of this method in the interface which we are implemented for.

File: store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+
+    @Override
+    public void flush() throws LogFailedException {
+        /* flush is just needed for Redolog Writes based on files (i. e. FileLogWriter)*/
+    }
+
+    @Override
+    public synchronized long getSize() {
+        if (conn == null) {
+            throw new RuntimeException("Redolog connection closed");
```
[the remark this answers is not in the document]
Comment (jnambo) [reply store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java]: This Exception has to be thrown this way if we don't want to change the semantic of this method in the interface which we are implemented for.

File: store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jnambo) [reply store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java]: We need to throw the exception this way if we don't want to change the LogWriter interface

File: store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jnambo) [reply store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java]: We need to throw the exception this way if we don't want to change the LogWriter interface

File: store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jnambo) [reply store/src/java/com/zimbra/cs/redolog/logger/DbLogWriter.java]: We need to throw the exception this way if we don't want to change the LogWriter interface

**Comment by fatkudu** [comment]
We are planning to treat these changes as a "done deal" for the HA branch for now and address issues, design problems as they come up via testing and further review.  I am still fundamentally unhappy with the decision to use the database as storage, but presumably we can change the backend reasonably easily at some point in the future to something designed for optimised append storage.

**Pull request closed without merging**